File: apps/mobile/modules/import-sources/ios/CodedError.swift
import Foundation

/// Pure-layer error carrying a registry code. The Expo module bindings map it
/// to an exception whose `code` crosses the bridge as `error.code` in JS.
public struct CodedError: Error, Equatable {
  public let code: String
  public let message: String

  public init(_ code: String, _ message: String = "") {
    self.code = code
    self.message = message
  }
}

File: apps/mobile/modules/import-sources/android/src/main/java/expo/modules/importsources/CodedError.kt
package expo.modules.importsources

import android.system.ErrnoException
import android.system.OsConstants

/**
 * Error carrying a stable code string. The Expo module bindings map it to a
 * CodedException so JS reads `error.code`.
 */
class CodedError(val code: String, message: String = "") : Exception(message)

/** Classifies a copy/stream throw to a stable reason code. */
internal fun mapIoError(e: Exception): CodedError {
  if (e is CodedError) return e
  val errno = (e.cause as? ErrnoException) ?: (e as? ErrnoException)
  if (errno?.errno == OsConstants.ENOSPC || e.message?.contains("No space left") == true) {
    return CodedError("not-enough-space", e.message ?: "")
  }
  if (e is SecurityException) return CodedError("permission-denied", e.message ?: "")
  return CodedError("io-error", e.message ?: "")
}
